lib/Sema/CSSimplify.cpp

```diff
@@ -10634,15 +10663,103 @@ ConstraintSystem::simplifyApplicableFnConstraint(
                               ? ConstraintKind::OperatorArgumentConversion
                               : ConstraintKind::ArgumentConversion);
 
+    auto *argumentsLoc = getConstraintLocator(
+        outerLocator.withPathElement(ConstraintLocator::ApplyArgument));
+
+    auto *argumentList = getArgumentList(argumentsLoc);
     // The argument type must be convertible to the input type.
     auto matchCallResult = ::matchCallArguments(
-        *this, func2, func1->getParams(), func2->getParams(), subKind,
-        outerLocator.withPathElement(ConstraintLocator::ApplyArgument),
-        trailingClosureMatching);
+        *this, func2, argumentList, func1->getParams(), func2->getParams(),
+        subKind, argumentsLoc, trailingClosureMatching);
 
     switch (matchCallResult) {
-    case SolutionKind::Error:
+    case SolutionKind::Error: {
+      auto resultTy = func2->getResult();
+
+      // If this is a call that constructs a callable type with
+      // trailing closure(s), closure(s) might not belong to
+      // the constructor but rather to implicit `callAsFunction`,
+      // there is no way to determine that without trying.
+      if (resultTy->isCallableNominalType(DC) &&
+          argumentList->hasAnyTrailingClosures()) {
+        auto *calleeLoc = getCalleeLocator(argumentsLoc);
+
+        bool isInit = false;
+        if (auto overload = findSelectedOverloadFor(calleeLoc)) {
+          isInit = bool(dyn_cast_or_null<ConstructorDecl>(
+              overload->choice.getDeclOrNull()));
+        }
+
+        if (!isInit)
+          return SolutionKind::Error;
+
+        auto &ctx = getASTContext();
+        auto numTrailing = argumentList->getNumTrailingClosures();
+
+        SmallVector<Argument, 4> newArguments;
+        SmallVector<Argument, 4> trailingClosures;
+
+        for (unsigned i = 0, n = argumentList->size(); i != n; ++i) {
+          if (argumentList->isTrailingClosureIndex(i)) {
+            trailingClosures.push_back(argumentList->get(i));
+          } else {
+            newArguments.push_back(argumentList->get(i));
+          }
+        }
```

Comment:
FWIW `ArgumentList` exposes iterator ranges for this e.g:

```cxx
SmallVector<Argument, 4> newArguments(argumentList->getNonTrailingArgs());
SmallVector<Argument, 4> trailingClosures(argumentList->getTrailingClosures());
```

Reply:
For some reason I thought that SmallVector would only accept begin+end and decided to just iterate instead…

Reply:
Adjusted, thank you!